flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/reader/ArrayPoolDataIteratorBatcher.java

```diff
@@ -0,0 +1,126 @@
+
+  private transient Pool<T[]> pool;
+
+  ArrayPoolDataIteratorBatcher(Configuration config, RecordFactory<T> recordFactory) {
+    this.batchSize = config.getInteger(FlinkConfigOptions.SOURCE_READER_FETCH_BATCH_RECORD_COUNT);
+    this.handoverQueueSize = config.getInteger(SourceReaderOptions.ELEMENT_QUEUE_CAPACITY);
+    this.recordFactory = recordFactory;
+  }
+
+  @Override
+  public CloseableIterator<RecordsWithSplitIds<RecordAndPosition<T>>> batch(
+      String splitId, DataIterator<T> inputIterator) {
+    // lazily create pool as it is not serializable
+    if (pool == null) {
+      pool = createPoolOfBatches(handoverQueueSize);
+    }
+    return new ArrayPoolBatchIterator(splitId, inputIterator, pool);
+  }
+
+  private Pool<T[]> createPoolOfBatches(int numBatches) {
+    Pool<T[]> poolOfBatches = new Pool<>(numBatches);
+    for (int batchId = 0; batchId < numBatches; batchId++) {
+      T[] batch = recordFactory.createBatch(batchSize);
+      poolOfBatches.add(batch);
+    }
+    return poolOfBatches;
+  }
+
+  private class ArrayPoolBatchIterator implements CloseableIterator<RecordsWithSplitIds<RecordAndPosition<T>>> {
+
+    private final String splitId;
+    private final DataIterator<T> inputIterator;
+    private final Pool<T[]> pool;
+
+    ArrayPoolBatchIterator(String splitId, DataIterator<T> inputIterator, Pool<T[]> pool) {
+      this.splitId = splitId;
+      this.inputIterator = inputIterator;
+      this.pool = pool;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return inputIterator.hasNext();
+    }
+
+    @Override
+    public RecordsWithSplitIds<RecordAndPosition<T>> next() {
+      T[] batch = getCachedEntry();
+      int recordCount = 0;
+      while (inputIterator.hasNext() && recordCount < batchSize) {
+        // The record produced by inputIterator can be reused like for the RowData case.
+        // inputIterator.next() can't be called again until the copy is made
+        // since the record is not consumed immediately.
+        T nextRecord = inputIterator.next();
+        recordFactory.clone(nextRecord, batch, recordCount);
+        recordCount++;
+        if (!inputIterator.currentFileHasNext()) {
+          // break early so that records in the ArrayResultIterator
+          // have the same fileOffset.
+          break;
+        }
+      }
+
+      if (recordCount == 0) {
+        return null;
+      } else {
+        return ArrayBatchRecords.forRecords(splitId, pool.recycler(), batch, recordCount,
+            inputIterator.fileOffset(), inputIterator.recordOffset() - recordCount);
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (inputIterator != null) {
+        inputIterator.close();
+      }
+    }
+
+    private T[] getCachedEntry() {
+      try {
+        return pool.pollEntry();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new UncheckedIOException(new IOException("Interrupted"));
```

Comment:
If it isn't an IOException, then we shouldn't create one to throw.

You could throw a RuntimeException here.

Reply:
will update to RuntimeException